lib/plugins/bull-mq-metrics/MetricsCollector.ts

```diff
@@ -0,0 +1,167 @@
+    labelNames: ['queue'],
+  }),
+  waitingGauge: new prometheus.Gauge({
+    name: `${prefix}_jobs_waiting`,
+    help: 'Total number of jobs waiting to be processed',
+    labelNames: ['queue'],
+  }),
+  processedDuration: new prometheus.Histogram({
+    name: `${prefix}_jobs_processed_duration`,
+    help: 'Processing time for completed jobs (processing until completed)',
+    buckets: histogramBuckets,
+    labelNames: ['queue'],
+  }),
+  completedDuration: new prometheus.Histogram({
+    name: `${prefix}_jobs_completed_duration`,
+    help: 'Completion time for jobs (created until completed)',
+    buckets: histogramBuckets,
+    labelNames: ['queue'],
+  }),
+})
+
+export class MetricsCollector {
+  private readonly options: MetricCollectorOptions
+  private readonly metrics: Metrics
+  private active = true
+
+  constructor(
+    private readonly redis: Redis,
+    private readonly registry: prometheus.Registry,
+    options: Partial<MetricCollectorOptions>,
+  ) {
+    this.options = {
+      bullMqPrefix: 'bull',
+      metricsPrefix: 'bullmq',
+      excludedQueues: [],
+      collectionIntervalInMs: 5000,
+      histogramBuckets: [20, 50, 150, 400, 1000, 3000, 8000, 22000, 60000, 150000],
+      queueDiscoverer: new BackgroundJobsBasedQueueDiscoverer(this.redis),
+      ...options,
+    }
+
+    this.metrics = this.registerMetrics(this.registry, this.options)
+  }
+
+  async start() {
+    const queueNames = await this.options.queueDiscoverer.discoverQueues()
+    await Promise.all(
+      queueNames.map((name) => this.observeQueue(name, this.redis, this.metrics, this.options)),
+    )
+  }
+
+  stop() {
+    this.active = false
+  }
+
+  private registerMetrics(
+    registry: prometheus.Registry,
+    { metricsPrefix, histogramBuckets }: MetricCollectorOptions,
+  ): Metrics {
+    const metrics = getMetrics(metricsPrefix, histogramBuckets)
+    const metricNames = Object.keys(metrics)
+
+    // If metrics are already registered, just return them to avoid triggering a Prometheus error
+    if (metricNames.length > 0 && registry.getSingleMetric(metricNames[0])) {
+      const retrievedMetrics = registry.getMetricsAsArray()
+      const returnValue: Record<string, prometheus.MetricObject> = {}
+
+      for (const metric of retrievedMetrics) {
+        if (metricNames.includes(metric.name)) {
+          returnValue[metric.name as keyof Metrics] = metric
+        }
+      }
+
+      return returnValue as unknown as Metrics
+    }
+
+    for (const metric of Object.values(metrics)) {
+      registry.registerMetric(metric)
+    }
+
+    return metrics
+  }
+
+  private async observeQueue(
+    name: string,
+    redis: Redis,
+    metrics: Metrics,
+    { collectionIntervalInMs }: MetricCollectorOptions,
+  ) {
+    const queue = new Queue(name, { connection: redis })
+    const events = new QueueEvents(name, { connection: redis })
+
+    // eslint-disable-next-line @typescript-eslint/no-misused-promises
```

Comment:
what triggered this error?

Reply:
```
134:28  error  Promise returned in function argument where a void return was expected  @typescript-eslint/no-misused-promises
```

Yeah I don't get this rule. Probably it is triggered by us awaiting something (a job) and if that place throws we are not catching the error. But I'm not exactly sure what's the expected code here.

Reply:
Or maybe the argument type implies a synchronous function but we are passing an `async` one - that is probably the correct error report.

I can fix it by wrapping the callback into `void` IIFE like that:

    ```
    events.on('completed', (completedJob: { jobId: string }) => {
      void (async () => {
        const job = await queue.getJob(completedJob.jobId)
        if (!job) {
          return
        }

        if (job.finishedOn) {
          metrics.completedDuration.labels({ queue: name }).observe(job.finishedOn - job.timestamp)

          if (job.processedOn) {
            metrics.processedDuration
              .labels({ queue: name })
              .observe(job.finishedOn - job.processedOn)
          }
        }
      })()
    })
    ```